### ISSUE TITLE
Map localized language names to ISO codes for CSL output

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#2699]
https://forums.zotero.org/discussion/comment/412045/#Comment_412045

**Comment by AbeJellinek** [comment]
I think that this is another thing we'll eventually want to visually indicate in the item box - a parsed language name should be displayed in the user's language when not being edited. Kind of leads to a weird situation where the value the user has entered, the value they see when not editing, and the value passed to CSL are all different, but I think it makes the most sense in the end.

[Intl.DisplayNames](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DisplayNames) can give us display names for languages in the user's locale. We could also use that to add more mappings: instead of just supporting English names ("Spanish") and the localized name in the target language ("español"), we could also support names for the language in other locales ("espagnol"). Is that worth trying?

**Comment by dstillman** [comment]
> we could also support names for the language in other locales ("espagnol")

Eh, going out of our way to translate every other language's name for every other language doesn't really seem worth it. I doubt this would really come up that much.

Also, @adam3smith or @bwiernik can correct me if I'm wrong, but I believe this field does approximately nothing at the moment other than [disable title-casing when set to something other than English](https://www.zotero.org/support/kb/preventing_title_casing_for_non-english_titles), though maybe that will change in the future. Naive parsing by the processor would also still work in most cases (e.g., "español" and "espagnol" would both be parsed as `es`).

The one thing that might be worth it is parsing known names into codes and actually storing the codes, which would then be displayed localized. One longstanding issue with our date parsing is that basing it on the current locale means that, in a group library, the same stored non-ISO date can be parsed in different ways for different people (!). We can avoid that here by just storing a cleanly parsing language name into the code. So if you enter "español", it will be parsed into `es`, you would see "Español" when the field closes, and then your French colleague will see "Espagnol". (I guess we're not worrying about definite articles.)

**Comment by bwiernik** [comment]
Yes, currently CSL only uses language for English casing controls. Jurism/CSL-m uses ISO code matching for per-locale bibliography formatting. That sort of multilingual support is something we are hoping to provide in a more modular way in a future CSL version.

**Comment by AbeJellinek** [comment]
How would that work with translation? Should we try to parse whatever the translator gives us, too, or just store that verbatim? If we stored it verbatim, I'm assuming that we'd localize it in the item box if it's an ISO code and show it as-is otherwise.

**Comment by bwiernik** [comment]
I think we should try to parse what the translator gives us

**Comment by dstillman** [comment]
Yeah, we can parse from a utility function called from`Item::setField()`, like the [date/accessDate stuff](https://github.com/zotero/zotero/blob/95239db0fd950b5701af900b29eb9c9a48c23ae4/chrome/content/zotero/xpcom/data/item.js#L806-L827).

(The similar [ISBN stuff](https://github.com/zotero/zotero/blob/95239db0fd950b5701af900b29eb9c9a48c23ae4/chrome/content/zotero/xpcom/data/item.js#L787-L803) is outside the `if (!loadIn)`, probably by mistake — I think that makes it apply to data coming from the DB as well, which would make it appear parsed but might prevent changing an existing stored value. Doesn't really matter if everyone has a client that shows the auto-corrected value, but then it wouldn't be possible to change and sync the corrected value to the online library, for example, so it's probably best to stay within `if (!loadIn)`. A better approach for existing values might be to apply these sorts of transformations in Update Metadata.)

**Comment by AbeJellinek** [comment]
Alright, see new PR.
